subsys/bluetooth/host/conn.c

```diff
@@ -1783,47 +1816,64 @@ void bt_conn_foreach(int type, void (*func)(struct bt_conn *conn, void *data),
 	int i;
 
 	for (i = 0; i < ARRAY_SIZE(acl_conns); i++) {
-		if (!atomic_get(&acl_conns[i].ref)) {
+		struct bt_conn *conn = bt_conn_ref(&acl_conns[i]);
+
+		if (!conn) {
 			continue;
 		}
 
-		if (!(acl_conns[i].type & type)) {
+		if (!(conn->type & type)) {
+			bt_conn_unref(conn);
 			continue;
 		}
 
-		func(&acl_conns[i], data);
+		func(conn, data);
+		bt_conn_unref(conn);
 	}
 #if defined(CONFIG_BT_BREDR)
 	if (type & BT_CONN_TYPE_SCO) {
 		for (i = 0; i < ARRAY_SIZE(sco_conns); i++) {
-			if (!atomic_get(&sco_conns[i].ref)) {
+			struct bt_conn *conn = bt_conn_ref(&sco_conns[i]);
+
+			if (!conn) {
 				continue;
 			}
 
-			func(&sco_conns[i], data);
+			func(conn, data);
+			bt_conn_unref(conn);
 		}
 	}
 #endif /* defined(CONFIG_BT_BREDR) */
 
 #if defined(CONFIG_BT_ISO)
 	if (type & BT_CONN_TYPE_ISO) {
 		for (i = 0; i < ARRAY_SIZE(iso_conns); i++) {
-			if (!atomic_get(&iso_conns[i].ref)) {
+			struct bt_conn *conn = bt_conn_ref(&iso_conns[i]);
+
+			if (!conn) {
 				continue;
 			}
 
-			func(&iso_conns[i], data);
+			func(conn, data);
+			bt_conn_unref(conn);
 		}
 	}
 #endif /* defined(CONFIG_BT_ISO) */
 }
 
 struct bt_conn *bt_conn_ref(struct bt_conn *conn)
 {
-	atomic_val_t old = atomic_inc(&conn->ref);
+	atomic_val_t old;
 
-	BT_DBG("handle %u ref %u -> %u", conn->handle, old,
-	       atomic_get(&conn->ref));
+	do {
```

Comment:
Can you please add a comment block here explaining the "algorithm" here?

Reply:
Also please document the new behavior in the Doxygen block in the header file.

Reply:
I added a note in the header already.

Added a comment block to explain the "algorithm".